### PR TITLE
Fix #450: support 64-bit indices in Louvain

### LINKED ISSUE #450: 64-bit Louvain? Support more than 2147483647 edges
* scikit-network version: 0.19.1
* Python version: 3.8.5
* Operating System: CentOS Linux 7 (Core)

### Description

Thanks for this great library! I'm trying to run `clustering.Louvain` on graphs with billions of edges (1 to 15 billion) and few millions of nodes (~5 million). The algorithm runs fine with 1,5 billion rows, but with more than that it fails without a traceback and exit code SIGSEV. Since max int is 2147483647, I believe the problem is `louvain.py` uses `np.int32` in multiple places and `louvain_core.pyx` uses `int`s instead of `long`s.

My question is: is it fine to open a PR increasing Louvain to deal with 64-bit integers? Or even leveraging from "fused types" and support both versions (see: https://stackoverflow.com/questions/28112599/writing-cython-for-both-32-bit-and-64-bit)?

### PR DESCRIPTION
**Pull request by fjsj**
### Modifications:
* Closes " 64-bit Louvain? Support more than 2147483647 edges #450"

### Impacted submodules:
 * sknetwork.clustering.louvain

### This pull request:
- [ ] **does not decrease code coverage**
- [ ] **passes the tests**
- [ ] **is documented** and **the documentation build passes**
- [ ] **has PEP8-compliant code** and **explicit variable naming**
- [ ] **has a tutorial** (facultative but recommended)

*Any doubts about some technicalities? Do not hesitate to look at the [dedicated Wiki](https://github.com/sknetwork-team/scikit-network/wiki/Contributing-guide).*

### CODE REVIEW

File: requirements_dev.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ pip==20.2.2
 # base dependencies
 numpy==1.19.1
 scipy==1.5.2
+psutil==5.7.3
```
Comment (fjsj) [inline requirements_dev.txt:7]: Is adding this dependency fine? [psutil](https://github.com/giampaolo/psutil) is the most reliable way to get the total RAM size in multiple platforms. 6.8k stars in GitHub.

File: sknetwork/clustering/louvain.py [hunk 2]
```diff
@@ -103,6 +104,12 @@ def __init__(self, resolution: float = 1, modularity: str = 'dugue', tol_optimiz
         self.n_aggregations = n_aggregations
         self.shuffle_nodes = shuffle_nodes
         self.random_state = check_random_state(random_state)
+        if psutil.virtual_memory().total >= 256 * (10 ** 9):
+            # Use int64 for node indices
+            # if there's more than 256 GBs of RAM available.
+            self.np_int_t = np.int64
```
Comment (fjsj) [inline sknetwork/clustering/louvain.py:110]: May we add a new parameter to the constructor to force the `np_int_t`?

File: sknetwork/clustering/louvain_core.pyx [hunk 3]
```diff
@@ -6,11 +6,14 @@ from libcpp.set cimport set
 from libcpp.vector cimport vector
 cimport cython
 
+ctypedef fused int_or_long:
```
Comment (fjsj) [inline sknetwork/clustering/louvain_core.pyx:9]: https://cython.readthedocs.io/en/latest/src/userguide/fusedtypes.html

**Comment by fjsj** [comment]
I'm having issues with psutil builds in Travis.

**Comment by nathandelara** [comment]
Thanks for the PR!

We are currently trying to think of ways not to use this extra dependency.
You are experimenting first hand the troubles that we usually encounter with new dependencies and continuous integration.
So adding one extra dependency just for this very specific feature might not be worth it.

In my opinion, there are still many workarounds to explore.
Hopefully, we will give you a feedback soon.

Best

**Comment by fjsj** [comment]
Sounds great @nathandelara.
I was able to build locally psutil in a Mac and CentOS 7 just fine. I suspect Travis may be making things more difficult here.
But I agree ideally a new dependency shouldn't be added here.

**Comment by QLutz** [comment]
Hello,

We gave this some more thought and found that the 32/64-bit index switch can be handled in a much simpler way: letting SciPy choose the most appropriate size and sticking to it. Using the fused type you introduced, it seems to work without much effort. I tried making a commit to add those changes but it seems your organisation's parameters won't allow me to push on your fork. If you can apply these changes yourself (they are minimal), they consist in:

  * removing `psutil`:
    * `setup.py`, line 24
    * `louvain.py`, line 13
  * removing the `np_int_t` attribute:
    * `louvain.py`, lines 107-112
    * `louvain.py`, line 198
  * adapting the tests in `test_louvain.py`:
    * Remove `from unittest import mock`
    * Change `test_options_with_64` to:
    ```python
    def test_options_with_64_bit(self):
        adjacency = karate_club()
        # force 64-bit index
        adjacency.indices = adjacency.indices.astype(np.int64)
        adjacency.indptr = adjacency.indptr.astype(np.int64)

        # resolution
        louvain = Louvain(resolution=2)
        labels = louvain.fit_transform(adjacency)
        self.assertEqual(len(set(labels)), 7)

        # tolerance
        louvain = Louvain(resolution=2, tol_aggregation=0.1)
        labels = louvain.fit_transform(adjacency)
        self.assertEqual(len(set(labels)), 12)

        # shuffling
        louvain = Louvain(resolution=2, shuffle_nodes=True, random_state=42)
        labels = louvain.fit_transform(adjacency)
        self.assertEqual(len(set(labels)), 9)

        # aggregate graph
        louvain = Louvain(return_aggregate=True)
        labels = louvain.fit_transform(adjacency)
        n_labels = len(set(labels))
        self.assertEqual(louvain.adjacency_.shape, (n_labels, n_labels))

        # aggregate graph
        Louvain(n_aggregations=1, sort_clusters=False).fit(adjacency)
    ```

    * In `test_options`, do not forget to remove:

    ```python
    # check used np.int type
    louvain = Louvain()
    self.assertEqual(louvain.np_int_t, np.int32)
    ```

**Comment by fjsj** [comment]
Sounds great, thanks for the input! I'll do these updates ASAP.

**Comment by fjsj** [comment]
OK, just force-pushed to have a single commit and added an additional check `self.assertEqual(labels.dtype, np.int64)`.
Please LMK if it's fine or not.

Thanks again for reviewing this!

**Comment by QLutz** [comment]
Hi,

The build failed due to unrelated issues that arose after you opened the PR. Can you merge with the latest commit of develop to correct them?

Thanks for the additional check! Also, if you happen to be able to test the algorithm on a graph that needs 64-bit indexing, do not hesitate to give us feedback (especially if there are any issues).

Thanks for contributing!

**Comment by fjsj** [comment]
Failed again, this time on Mac setup at "Installing iTunes, iTunes Device Support Update, Security Update 2020-005, Safari, Command Line Tools (macOS High Sierra version 10.13) for Xcode".

Tested on my local Mac, working fine.

**Comment by QLutz** [comment]
I've tried to make a lighter update for Travis not to timeout. Could you give it a try?

We're looking forward to hearing from you then!

**Comment by QLutz** [comment]
Closing this now but do not hesitate to tell us how the package behaves on huge graphs.

Thanks a lot for the contribution!
